crates/rustc_codegen_spirv/src/builder/builder_methods.rs

```diff
@@ -1974,6 +1975,128 @@ impl<'a, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'tcx> {
+                    std::iter::once(Operand::LiteralInt32(0)),
+                );
+                drop(emit_global);
+
+                let bindless_idx = args[1].def(self);
+                let offset_arg = args[2].def(self);
+                let mut composite_components = vec![];
+
+                let two = self.constant_int(uint_ty, 2).def(self);
+
+                let dword_offset = self
+                    .emit()
+                    .shift_right_arithmetic(uint_ty, None, offset_arg, two)
+                    .unwrap();
+
+                match &data {
+                    &SpirvType::Adt {
+                        ref field_types,
+                        ref field_offsets,
+                        ..
+                    } => {
+                        for (ty, offset) in field_types.iter().zip(field_offsets.iter()) {
+                            // assert!(offset % 4 == 0, "Field elements need to be 4 byte aligned");
+                            // jb-todo: need to check size is a multiple of 4 as well
+                            // jb-todo: need to handle for eg. 64-bit types
+                            // jb-todo: need to handle vector types and nested ADTs
+                            let offset = if offset > &Size::ZERO {
+                                let element_offset = self
+                                    .constant_int(uint_ty, offset.bytes() as u64 / 4)
+                                    .def(self);
+
+                                self.emit()
+                                    .i_add(uint_ty, None, dword_offset, element_offset)
+                                    .unwrap()
+                            } else {
+                                dword_offset
+                            };
+
+                            let indices = vec![bindless_idx, zero, offset];
+
+                            let result = self
+                                .emit()
+                                .access_chain(uniform_uint_ptr, None, buffer, indices)
+                                .unwrap();
+
+                            let load_res = self
+                                .emit()
+                                .load(uint_ty, None, result, None, std::iter::empty())
+                                .unwrap();
+
+                            let bitcast_res = self.emit().bitcast(*ty, None, load_res).unwrap();
+
+                            composite_components.push(bitcast_res as u32);
+                        }
+
+                        let adt = data.def(rustc_span::DUMMY_SP, self);
+
+                        let constructed_adt = self
+                            .emit()
+                            .composite_construct(adt, None, composite_components)
+                            .unwrap()
+                            .with_type(adt);
+
+                        return self.store(constructed_adt, args[0], Align::from_bytes(0).unwrap());
+                    }
+                    _ => {}
```

Comment:
There are quite a few more cases to handle here: scalars / vectors, also nested types etc.